system/include/emscripten/wire.h

```diff
@@ -15,22 +15,66 @@
 #define EMSCRIPTEN_ALWAYS_INLINE __attribute__((always_inline))
 
 namespace emscripten {
+    #ifndef EMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES
+    #define EMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES 1
+    #endif
+
+
+    #if EMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES
+    constexpr bool has_unbound_type_names = true;
+    #else
+    constexpr bool has_unbound_type_names = false;
+    #endif
+
     namespace internal {
         typedef void (*GenericFunction)();
 
-        typedef const struct _TYPEID* TYPEID;
+        typedef const struct _TYPEID {}* TYPEID;
+
+
+        // We don't need the full std::type_info implementation.  We
+        // just need a unique identifier per type and polymorphic type
+        // identification.
+        
+        template<typename T>
+        struct CanonicalizedID {
+            static TYPEID get() {
+                static _TYPEID c;
+                return &c;
+            }
+        };
+
+        template<typename T>
+        struct Canonicalized {
+            typedef typename std::remove_cv<typename std::remove_reference<T>::type>::type type;
+        };
+
+        template<typename T>
+        struct LightTypeID {
+            static TYPEID get() {
+                typedef typename Canonicalized<T>::type C;
+                if (has_unbound_type_names || std::is_polymorphic<C>::value) {
+                    return reinterpret_cast<TYPEID>(&typeid(C));
+                } else {
+                    return CanonicalizedID<C>::get();
+                }
+            }
+        };
+
+        template<typename T>
+        const TYPEID getLightTypeID(const T& value) {
+            typedef typename Canonicalized<T>::type C;
+            if (has_unbound_type_names || std::is_polymorphic<C>::value) {
+                return reinterpret_cast<TYPEID>(&typeid(value));
+            } else {
+                return LightTypeID<T>::get();
+            }
+        }
 
-        // This implementation is technically not legal, as it's not
-        // required that two calls to typeid produce the same exact
-        // std::type_info instance.  That said, it's likely to work
-        // given Emscripten compiles everything into one binary.
-        // Should it not work in the future: replace TypeID with an
-        // int, and store all TypeInfo we see in a map, allocating new
-        // TypeIDs as we add new items to the map.
```

Comment:
Is this comment not relevant anymore? We're seeing this exact issue for shared libraries in https://github.com/emscripten-core/emscripten/issues/16711

Or does CanonicalizedID handle this case, and if we switch to EMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES it should work?